File: 06/c06.js
#!/usr/bin/env node

var net = require('net');
var util = require('util');
var crypto = require('crypto');

var CLIENT = "CLIENT";
var SERVER = "SERVER";
var ERROR = "Error";

var KEY_MSG = "key";
var KEYPHRASE_MSG = "keyphrase";
var RESULT_MSG = "result";

var options = {
	'port': 6969,
	'host': "54.83.207.90",
};


process.stdin.resume();
process.stdin.setEncoding('utf8');
var payload = null;
process.stdin.on('data', function (chunk) {
	payload = chunk.trim();
});

var socket = net.connect(options);

socket.on('data', function(data) {
	data = data.toString().trim();
	if(data == ERROR) {
		socket.end();
		return;
	}

	// Parses msg
	data = data.split(':');
	var msg = data[1];
	var dmsg = msg.split('|'); 
	data = data[0].split('->');
	var from = data[0];
	var to = data[1];
	
	// Man in the middle
	if(from == CLIENT && dmsg[0] == KEY_MSG) {
		sim_c_dh = crypto.createDiffieHellman(dmsg[1], 'hex');
		sim_c_dh.generateKeys();
		sim_c_secret = sim_c_dh.computeSecret(dmsg[2], 'hex');

		sim_s_dh = crypto.createDiffieHellman(256);
		sim_s_dh.generateKeys();
		msg = util.format('key|%s|%s\n', sim_s_dh.getPrime('hex'), sim_s_dh.getPublicKey('hex'));

	} else if(from == SERVER && dmsg[0] == KEY_MSG) {
		sim_s_secret = sim_s_dh.computeSecret(dmsg[1], 'hex');
		msg = util.format('key|%s\n', sim_c_dh.getPublicKey('hex'));

	} else if(from == CLIENT && dmsg[0] == KEYPHRASE_MSG) {
		var cipher = crypto.createCipheriv('aes-256-ecb', sim_s_secret, '');
		var keyphrase1 = cipher.update(payload, 'utf8', 'hex') + cipher.final('hex');
		msg = util.format('keyphrase|%s\n', keyphrase1);

		//var decipher2 = crypto.createDecipheriv('aes-256-ecb', sim_c_secret, '');
		//var keyphrase3 = decipher2.update(dmsg[1], 'hex', 'utf8') + decipher2.final('utf8');
		//process.stdout.write("Original keyphrase: " + keyphrase3 + '\n');
	} else if(from == SERVER && dmsg[0] == RESULT_MSG) {
		var decipher = crypto.createDecipheriv('aes-256-ecb', sim_s_secret, '');
		var keyphrase2 = decipher.update(dmsg[1], 'hex', 'utf8') + decipher.final('utf8');

		var cipher2 = crypto.createCipheriv('aes-256-ecb', sim_c_secret, '');
		var result = cipher2.update(keyphrase2, 'utf8', 'hex') + cipher2.final('hex');
		msg = util.format('result|%s\n', result);

		// prints solution
		process.stdout.write(keyphrase2 + '\n');
	}

	//process.stdout.write("------------->:" + msg + "\n");
	socket.write(msg);
});

File: 06/c06.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

#######################################
# Tuenti Challange 4
# Problem 6
#
# Angel Calvo
#######################################

# DO NOT WORK!

require "socket"
require "openssl"

HOST = "54.83.207.90"
PORT = 6969

EXPR = /^(\w+)->(\w+):(.+)$/
CLIENT = "CLIENT"
SERVER = "SERVER"
ERROR = "Error"

KEY_MSG = "key"
KEYPHRASE_MSG = "keyphrase"
RESULT_MSG = "result"

def encrypt(data, key, iv = nil)
  aes = OpenSSL::Cipher::Cipher.new("AES-256-ECB")
  aes.encrypt
  aes.key = key
  aes.iv = iv if iv != nil
  aes.update(data) + aes.final
end

def decrypt(encrypted_data, key, iv = nil)
    aes = OpenSSL::Cipher::Cipher.new("AES-256-ECB")
    aes.decrypt
    aes.key = key
    aes.iv = iv if iv != nil
    aes.update(encrypted_data) + aes.final  
  end
  
# Main block
if __FILE__ == $0
  socket = TCPSocket.new HOST, PORT
  payload = ARGF.readline.chomp
  
  # DF
  s_dh = OpenSSL::PKey::DH.new(256) 
  s_dh.generate_key!
  #s_dh = OpenSSL::PKey::DH.new(c_dh.public_key.to_der)
  #s_dh.generate_key!
  #c_symm_key = c_dh.compute_key(c_dh.pub_key)
  #s_symm_key = s_dh.compute_key(s_dh.pub_key)

  while line = socket.gets
	puts line.chomp
	exit() if line.chomp == ERROR

	# Parses package
	from, to, msg = line.chomp.match(EXPR).captures
	dmsg = msg.split("|")

 	if from == CLIENT and dmsg[0] == KEY_MSG
	  c_dh = OpenSSL::PKey::DH.new([dmsg[1]].pack('H*'))
	  c_dh.generate_key!
	  c_key = c_dh.compute_key(dmsg[2])
	  
	  msg = "#{KEY_MSG}|#{s_dh.public_key.p.to_s(16)}|#{s_dh.public_key.to_der.unpack('H*')[0]}"
	  puts "#{dmsg[1]}"
	elsif from == SERVER and dmsg[0] == KEY_MSG
	  s_key = s_dh.compute_key(dmesg[1])
	  msg = "#{KEY_MSG}|#{s_dh.pub_key.to_s(16)}"
	  
 	elsif from == CLIENT and dmsg[0] == KEYPHRASE_MSG
	  puts "decrypt(#{dmsg[1]}, [#{c_symm_key.unpack('H*')[0]}, #{s_symm_key.unpack('H*')[0]}, #{c_dh.pub_key.to_s(16)}, #{c_dh.priv_key.to_s(16)}, #{s_dh.pub_key.to_s(16)}, #{s_dh.priv_key.to_s(16)}])"
	  #puts decrypt([dmsg[1]].pack('H*'), c_symm_key.to_s).unpack('H*')[0]
	  c_payload = encrypt(payload, [c_dh.priv_key.to_s].pack('H*')).unpack('H*')[0]
 	  msg = "#{KEYPHRASE_MSG}|#{c_payload}"
	  
	elsif from == SERVER and dmsg[0] == RESULT_MSG
	  secret = decrypt(dmsg[1])
	  puts secret
 	end
	
	# Send
	puts "------------->:#{msg}"
	socket.puts "#{msg}\n"
  end
  socket.close  
end
